### PR TITLE
Add each methods

### PR DESCRIPTION
**Pull request by labpak**

### CODE REVIEW

File: lib/matrix/matrix.rb [hunk 1]
```diff
@@ -57,12 +61,54 @@ def self.convert(matrix)
...
-      each_with_index { |elem, _, _| block.call(elem) }
-      self
-    end
+     def each(which = :all)
+        return to_enum :each, which unless block_given?
+        last = column_count - 1
+        case which
+        when :all
+          block = Proc.new
+          rows.each do |row|
```
Comment (RobolabGs2) [inline lib/matrix/matrix.rb]: `rows` is slow, becouse make a copy. I think we need a simple cycle here.
@SID37, what do you think of it?

Comment (SID37) [reply lib/matrix/matrix.rb]: I agree

File: lib/matrix/matrix.rb [hunk 2]
```diff
@@ -57,13 +61,68 @@ def self.convert(matrix)
...
+              yield self[i, j]
+            end
+          end
+        when :diagonal
+          (0...row_count).each do |i|
+            (0...column_count).each do |j|
+              if i == j
+                yield self[i, j]
+              end
+            end
```
Comment (SID37) [inline lib/matrix/matrix.rb]: Now think about it again.

**Review by SID37 — CHANGES_REQUESTED** [review]

**Review by RobolabGs2 — APPROVED** [review]

**Review by SID37 — APPROVED** [review]

File: test/matrix/matrix_test.rb [hunk 3]
```diff
@@ -153,5 +153,111 @@ def test_not_zero
...
+    end
+
+    def test_each_upper_sq
+      m = Matrix[[1, 2], [3, 4]]
+      assert_equal [1, 2, 4], m.each(:upper).to_a
+    end
+
+    def test_each_upper_rec1
+      m = Matrix[[1, 2], [3, 4], [5, 6]]
+      assert_equal [1, 2, 4], m.each(:upper).to_a,
```
Comment (RobolabGs2) [inline test/matrix/matrix_test.rb]: Extra comma at end of line

**Review by RobolabGs2 — CHANGES_REQUESTED** [review]

**Review by RobolabGs2 — APPROVED** [review]

**Pull request merged**